File: code/894.py
'''
Author: lmio 2091319361@qq.com
Date: 2024-04-02 18:47:35
LastEditors: lmio 2091319361@qq.com
Description: 894. 所有可能的真二叉树
'''


from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

MX = 11
dp = [[] for _ in range(MX)]
dp[1] = [TreeNode()]
for i in range(2, MX):  # 计算 dp[i]
    dp[i] = [TreeNode(0, left, right)
            for j in range(1, i)  # 枚举左子树叶子数
            for left in dp[j]  #  枚举左子树
            for right in dp[i - j]]  # 枚举右子树


class Solution:
    def allPossibleFBT(self, n: int) -> List[Optional[TreeNode]]:
        return dp[n//2+1] if n&1 else []

File: code_cpp/894.cpp
/*
 * @Author: lmio 2091319361@qq.com
 * @Date: 2024-04-02 14:32:42
 * @LastEditors: lmio 2091319361@qq.com
 * @Description: 894. 所有可能的真二叉树
 */
#include<bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

vector<TreeNode*> dp[11];
auto init = [] {
    dp[1] = {new TreeNode()};
    for (int i = 2; i < 11; i++) {
        for (int j = 1; j < i; j++) {
            for (auto left: dp[j]) {
                for (auto right: dp[i-j]) {
                    dp[i].push_back(new TreeNode(0, left, right));
                }
            }
        }
    }
    return 0;
}();

class Solution {
public:
    vector<TreeNode*> allPossibleFBT(int n) {
        return dp[n&1 == 0? 0: n/2+1];
    }
};